File: client-android/lib/src/main/kotlin/stasis/client_android/lib/telemetry/analytics/AnalyticsCollector.kt
package stasis.client_android.lib.telemetry.analytics

import stasis.client_android.lib.telemetry.ApplicationInformation
import stasis.client_android.lib.utils.Try

interface AnalyticsCollector {
    fun recordEvent(name: String): Unit =
        recordEvent(name = name, attributes = emptyMap())

    fun recordEvent(name: String, result: Try<*>): Unit =
        recordEvent(name = name, "result" to if (result.isSuccess) "success" else "failure")

    fun recordEvent(name: String, vararg attributes: Pair<String, String>): Unit =
        recordEvent(name = name, attributes = attributes.toMap())

    fun recordFailure(e: Throwable): Unit = recordFailure(message = "${e.javaClass.simpleName} - ${e.message}")

    fun recordEvent(name: String, attributes: Map<String, String>)

    fun recordFailure(message: String)

    fun state(): Try<AnalyticsEntry>

    fun send()

    val persistence: AnalyticsPersistence?

    object NoOp : AnalyticsCollector {
        override fun recordEvent(name: String, attributes: Map<String, String>) = Unit

        override fun recordFailure(message: String) = Unit

        override fun state(): Try<AnalyticsEntry> =
            Try.Success(AnalyticsEntry.collected(app = ApplicationInformation.none()))

        override fun send() = Unit

        override val persistence: AnalyticsPersistence? = null
    }
}

File: layers/src/main/scala/stasis/layers/telemetry/analytics/AnalyticsCollector.scala
package stasis.layers.telemetry.analytics

import scala.concurrent.Future

import stasis.layers.telemetry.ApplicationInformation

trait AnalyticsCollector {
  def recordEvent(name: String): Unit =
    recordEvent(name = name, attributes = Map.empty[String, String])

  def recordEvent(name: String, attributes: (String, String)*): Unit =
    recordEvent(name = name, attributes = attributes.toMap)

  def recordFailure(e: Throwable): Unit = recordFailure(message = s"${e.getClass.getSimpleName} - ${e.getMessage}")

  def recordEvent(name: String, attributes: Map[String, String]): Unit

  def recordFailure(message: String): Unit

  def state: Future[AnalyticsEntry]

  def send(): Unit

  def persistence: Option[AnalyticsPersistence]
}

object AnalyticsCollector {
  object NoOp extends AnalyticsCollector {
    override def recordEvent(name: String, attributes: Map[String, String]): Unit = ()
    override def recordFailure(message: String): Unit = ()
    override def state: Future[AnalyticsEntry] = Future.successful(AnalyticsEntry.collected(app = ApplicationInformation.none))
    override def send(): Unit = ()
    override def persistence: Option[AnalyticsPersistence] = None
  }
}
